google-cloud-notification/src/main/java/com/google/cloud/notification/NotificationInfo.java

```diff
@@ -0,0 +1,371 @@
+/*
+ * Copyright 2015 Google Inc. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *       http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.google.cloud.notification;
+
+import static com.google.common.base.Preconditions.checkNotNull;
+
+import com.google.api.services.storage.model.Notification;
+import com.google.common.base.Function;
+import com.google.common.base.MoreObjects;
+import com.google.common.collect.ImmutableList;
+import com.google.common.collect.ImmutableMap;
+
+import com.google.pubsub.v1.TopicName;
+import java.io.Serializable;
+import java.util.List;
+import java.util.Map;
+import java.util.Objects;
+
+/**
+ * Google Storage Notification metadata;
+ *
+ * @see <a href="https://cloud.google.com/storage/docs/concepts-techniques#concepts">Concepts and
+ *      Terminology</a>
+ */
+public class NotificationInfo implements Serializable {
```

Comment:
This class looks immutable, do we want to use AutoValue?

Reply:
I'd like to keep the underlying collection fields as ImmutableLists (and -Sets and -Maps), but I don't want to expose guava stuff on the surface because this is a public class in a public repo. Is there a way to use Autovalue where the user can pass in a List to the Builder, but then the underlying object turns that into an ImmutableList?

Reply:
It's a little sneaky, but I think it's worth not having to write the whole class. Up to you.

```java
// AutoValue will override this.
// It's package-private; not part of surface.
abstract Builder setMapInternal(ImmutableMap<K,V> m);

// Users use this
public Builder setMap(Map<K,V> m) {
  return setMapInternal(ImmutableMap.copyOf(m));
}

// and do it the opposite way for NotificationInfo
```

Also, I didn't notice this before, but why did you need an abstract builder in the current code?

Reply:
Hmm I couldn't make my IDE happy with the AutoValue pattern.

I removed the abstract builder and replaced it with a concrete one.